GitUI/Theming/ThemeCssLoader.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Drawing;
+using System.IO;
+using System.Linq;
+using ExCSS;
+using GitCommands;
+using GitExtUtils.GitUI.Theming;
+using ResourceManager;
+
+namespace GitUI.Theming
+{
+    internal class ThemeCssLoader
+    {
+        private const string ClassSelector = ".";
+        private const string ColorProperty = "color";
+
+        private readonly TranslationString _invalidRule = new TranslationString("Invalid rule: {0}");
+        private readonly TranslationString _failedToLoadThemeFrom = new TranslationString("Failed to load theme from {0}");
+        private readonly TranslationString _fileNotFound = new TranslationString("File not found");
+        private readonly TranslationString _fileTooLarge = new TranslationString("File too large");
+
+        private readonly Parser _parser;
+        private readonly ICssUrlResolver _urlResolver;
+        private readonly string[] _allowedClasses;
+
+        private readonly Dictionary<AppColor, Color> _appColors = new Dictionary<AppColor, Color>();
+        private readonly Dictionary<KnownColor, Color> _sysColors = new Dictionary<KnownColor, Color>();
+        private readonly Dictionary<string, int> _specificityByColor = new Dictionary<string, int>();
+
+        private bool _parseCalled;
+
+        public ThemeCssLoader(ICssUrlResolver urlResolver, string[] allowedClasses)
+        {
+            _parser = new Parser();
+            _urlResolver = urlResolver;
+            _allowedClasses = allowedClasses;
+        }
+
+        public IReadOnlyDictionary<AppColor, Color> AppColors => _appColors;
+
+        public IReadOnlyDictionary<KnownColor, Color> SysColors => _sysColors;
+
+        public bool TryLoadCss(string filePath)
+        {
+            if (_parseCalled)
+            {
+                throw new InvalidOperationException($"{nameof(ThemeCssLoader)} only supports 1 call to {nameof(TryLoadCss)}");
+            }
+
+            _parseCalled = true;
+
+            return TryLoadCssImpl(filePath, cssImportChain: Array.Empty<string>());
+        }
+
+        private bool TryLoadCssImpl(string filePath, string[] cssImportChain)
```

Comment:
Since all failures appear to be terminal, I think there is little value in "TryXxx" methods, and we may as well throw all the way through the stack. This way the code would be terse, and we'd be able to bubble errors up to the UI layer, and show error messages to the user to action.

Reply:
ok